javascript/photo.js

```diff
@@ -115,64 +115,65 @@ backgrounds.Photo = new Model({
+        // * count < STARTUP_IMAGE_COUNT: We are yet to finish the quota of offline images before we can use images from the cache.
+        // * count >= STARTUP_IMAGE_COUNT but cachedImages.length == 0: Can happen when HTTP requests to r/EarthPorn fail. We fall back to display the backup images.
+        if (count < this.STARTUP_IMAGE_COUNT || cachedImages.length == 0) {
+            var order_index = count % this.BACKUP_IMAGES_COUNT;
+            this.displayImage(this.offlineImages[order[order_index]]);
         } else {
-            // TODO: Use variable cacheReadPos & cacheWritePos to prevent errors when fetch fails
-            var cachedImageNumber = cacheIndex - this.STARTUP_IMAGE_COUNT;
-            if (cachedImageNumber < 0) {
-                cachedImageNumber += cacheSize;
-            }
-            var cachedImages = JSON.parse(localStorage.getItem("cachedImages"));
-            this.displayImage(cachedImages[cachedImageNumber]);
+            this.displayImage(cachedImages[0]);
+            // Once we consume an image from the cache, remove it from the cache.
+            cachedImages.splice(0, 1);
+            localStorage.setItem("cachedImages", JSON.stringify(cachedImages));
         }
         var xmlHttp = new XMLHttpRequest();
         var parent = this;
         xmlHttp.onreadystatechange = function() {
+            // If the HTTP request is successful, then we store the image received into our cache. If the request is unsuccessful, the cache simply does not get updated.
             if (xmlHttp.readyState == 4 && xmlHttp.status == 200) {
                 var response = JSON.parse(xmlHttp.response);
                 // Fetch an image and cache it so it can be displayed quickly
                 var imageData = parent.parseImage(response, count);
-                var cachedImages = JSON.parse(localStorage.getItem("cachedImages"));
-                //var cacheIndex = Number(localStorage.getItem("cacheIndex"));
-                cachedImages[cacheIndex] = imageData;
+                cachedImages.push(imageData);
                 localStorage.setItem("cachedImages", JSON.stringify(cachedImages));
-            } else if (xmlHttp.readyState == 4) { // This can probably be replaced with an else
-                // HTTP request completed but was not successful
-                // if count < 3, then the background image is already set
-                if (count >= parent.STARTUP_IMAGE_COUNT) {
-                    var imageNumber = count % parent.BACKUP_IMAGES_COUNT;
-                    displayImage(parent.offlineImages[imageNumber]);
-                }
             }
         }
-        // TODO: Think about not calling the API after we get all images (or caching all images & then not calling?)
-        xmlHttp.open("GET", "https://www.reddit.com/r/EarthPorn/top/.json?limit=" + Number(numberOfImages), true);
-        xmlHttp.send(null);
+        // We send an HTTP request only if our cache is not full.
```

Comment:
When will it be equal? Assume cacheSize > 1, then each time we add to the cache, on the next call, we will pull from it.

Reply:
If the startup image count exceeds the cacheSize, then the cache will become full before we can use it.